### PR TITLE
Add split Go SDK

### PR DESCRIPTION
**Pull request by danielrbradley**
1. Generate secondary go SDK into `sdk/pulumi-azure-native-sdk` to match new [pulumi-azure-native-sdk repository](https://github.com/pulumi/pulumi-azure-native-sdk)
2. Insert root and leaf `go.mod` files as part of codegen (based on directories with `init.go` files).
3. Include replace directives so the modules can be used locally.
4. Incude current version number as part of go.mod references
    - Fall back to `latest` if not available.
    - Strip build number from codegen version to make compatible with go module versioning.
 5. Build each go module using find.
 6. Add `go_prepublish` target to strip `go.sum` files and `go.mod` replace directives
 7. Adapt GitHub workflows to generate and persist new go SDK between steps
 8. During publish, checkout new repository in sub-folder then tag and push

Outstanding issues:
- [Generated version utility assumes the SDK is in the same repository](https://github.com/pulumi/pulumi-azure-native-sdk/blob/v1.76.0-alpha.24984ad3/pulumiUtilities.go#L70) so won't get the exact correct version.

# Original investigation notes

Investigation into solving #1930 as a proof of concept and to decide a route forward.

Either approach below would be best combined with committing the new go SDK into a new repo such as `pulumi-azure-sdk` as ~200 tags will need to be added on each release, and it adds significantly more code.

## Add go.mod files to the original structure
Pros: The core generation is almost identical, just with extra files appended

Have to reference the root sdk module of the same version. This would require publishing the root SDK first with the new version number, then pulling in the new version number into the per-module go.mod then publishing each after.

We really need the provider to be included in each sub-module directly to remove any internal dependencies between modules that have to be published together.

## Split the schema and then generate independently

This results in every module having its own provider class (and config). This doesn't affect when using the default provider, but would look a little more messy with explicit providers.

Split by modifying discovery glob - makes discovery really quick for single modules.

Go package name is calculated from end of importBasePath in go language config.

`authorization`, `storate` and `synapse` seem to be mixins - not from specs. Could add options to skip adding these, have just filtered them out by token for now.

Generate the go.mod and add to list of files to emit.

### Optional: Remove Double Nesting

Go module is at something like `github.com/pulumi/pulumi-azure-native-sdk/compute`. Codegen wants to create a second `compute` folder within that, but that's not great to have to import `github.com/pulumi/pulumi-azure-native-sdk/compute/compute`.

How do we avoid the double nesting of module name in Go when there's only one module? Could manually change the paths for now. Should we just live with double nesting for now?

There's separate init files in the root and module folders. The root file has construct provider and registers resource package. The module file has the resource "construct" method and registers resource module. Seems like we can just rename the module init file on moving, then fix the import used for the PkgVersion.

## Opportunities to improve core codegen

- Add option to generate go.mod
- Allow option for a "default" module to be pulled to the top-level rather than every module having its own folder.

## Out of scope

- Integration of the new Go SDK for ProgramTest - could continue to generate the monolithic SDK temporarily for testing.
- Publishing workflow - would need to add steps to commit to a separate repo and add relevant tags.

### CODE REVIEW

**Comment by Frassle** [comment]
> Add option to generate go.mod

I think we need that in platform for some things anyway. We want to generate some types into the core sdk, but just the code files none of the project related stuff (i.e. no .csproj, go.mod, etc). I think that option would work for you here as well.

File: provider/cmd/pulumi-gen-azure-native/main.go [hunk 1]
```diff
@@ -271,10 +276,85 @@ func emitPackage(pkgSpec *schema.PackageSpec, language, outDir string) error {
 		return errors.Wrapf(err, "generating %s package", language)
 	}
 
+	// go-split ... add go.mods
```
Comment (danielrbradley) [inline provider/cmd/pulumi-gen-azure-native/main.go:279]: Remove this

File: provider/cmd/pulumi-gen-azure-native/main.go [hunk 2]
```diff
@@ -271,10 +276,85 @@ func emitPackage(pkgSpec *schema.PackageSpec, language, outDir string) error {
...
+	files["pulumi-azure-native-sdk/go.mod"] = []byte(`
+module github.com/pulumi/pulumi-azure-native-sdk
+
+go 1.17
+
+require (
+	github.com/blang/semver v3.5.1+incompatible
+	github.com/pkg/errors v0.9.1
+	github.com/pulumi/pulumi/sdk/v3 v3.37.2
```
Comment (danielrbradley) [inline provider/cmd/pulumi-gen-azure-native/main.go:312]: Parameterise versions

File: provider/cmd/pulumi-gen-azure-native/main.go [hunk 3]
```diff
@@ -271,10 +276,85 @@ func emitPackage(pkgSpec *schema.PackageSpec, language, outDir string) error {
...
+		version = "v" + buildVersion.String()
+	} else {
+		version = "latest"
+	}
 	for f, contents := range files {
 		if err := emitFile(outDir, f, contents); err != nil {
 			return errors.Wrapf(err, "emitting file %v", f)
 		}
+
+		if matches := re.FindStringSubmatch(f); len(matches) > 0 {
```
Comment (danielrbradley) [inline provider/cmd/pulumi-gen-azure-native/main.go:335]: Might be able to use https://pkg.go.dev/path/filepath

File: Makefile [hunk 4]
```diff
@@ -173,13 +173,18 @@ build_dotnet::
...
 generate_go:: .version
-	$(WORKING_DIR)/bin/$(CODEGEN) go $(shell cat .version)
+	$(WORKING_DIR)/bin/$(CODEGEN) go,go-split $(shell cat .version)
 	cd ${PACKDIR}/go/ && find . -type f -exec sed -i '' -e '/^\/\/.*/g' {} \;
 
 build_go::
 	# Only building the top level packages and building 1 package at a time to avoid OOMing
 	cd sdk/ && \
 	GOGC=50 go list github.com/pulumi/pulumi-azure-native/sdk/go/azure/... | grep -v "latest\|\/v.*"$ | xargs -L 1 go build
+	find sdk/pulumi-azure-native-sdk -type d -maxdepth 1 -exec sh -c "cd \"{}\" && go mod tidy && go build" \;
```
Comment (danielrbradley) [inline Makefile:183]: Need to address missing default versions

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -298,6 +307,15 @@ jobs:
...
         github.workspace}}/sdk/${{ matrix.language}}
+    - name: Tar split Go SDK folder
+      if: ${{ matrix.language == 'go' }}
+      run: tar -zcf sdk/pulumi-azure-native-sdk.tar.gz -C sdk/pulumi-azure-native-sdk .
+    - name: Upload split Go artifacts
+      uses: actions/upload-artifact@v2
+      if: ${{ matrix.language == 'go' }}
+      with:
+        name: pulumi-azure-native-sdk.tar.gz
+        path: ${{ github.workspace}}/sdk/pulumi-azure-native-sdk.tar.gz
```
Comment (danielrbradley) [inline .github/workflows/build.yml:318]: This should be untar

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -450,6 +468,14 @@ jobs:
     - name: Uncompress nodejs SDK
       run: tar -zxf ${{github.workspace}}/sdk/nodejs.tar.gz -C
         ${{github.workspace}}/sdk/nodejs
+    - name: Download go SDK
+      uses: actions/download-artifact@v2
+      with:
+        name: nodejs-sdk.tar.gz
+        path: ${{ github.workspace}}/sdk/
```
Comment (danielrbradley) [inline .github/workflows/build.yml:475]: Wrong asset

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -448,12 +470,30 @@ jobs:
...
       run: tar -zxf ${{github.workspace}}/sdk/nodejs.tar.gz -C
         ${{github.workspace}}/sdk/nodejs
+    - name: Download go SDK
+      uses: actions/download-artifact@v2
+      with:
+        name: nodejs-sdk.tar.gz
+        path: ${{ github.workspace}}/sdk/
+    - name: Uncompress go SDK
+      run: tar -zxf ${{github.workspace}}/sdk/nodejs.tar.gz -C
+        ${{github.workspace}}/sdk/nodejs
```
Comment (danielrbradley) [inline .github/workflows/release.yml:480]: Wrong SDK

**Review by danielrbradley — COMMENTED** [review]
Pair review with @thomas11 

Document & changelog

File: Makefile (hunk 4 above)
Comment (danielrbradley) [reply Makefile:201]: Resolved via #1949

**Review by viveklak — COMMENTED** [review]
Should/can we add an acceptance test for this?

File: provider/cmd/pulumi-gen-azure-native/main.go [hunk 8]
```diff
@@ -249,6 +255,95 @@ func emitPackage(pkgSpec *schema.PackageSpec, language, outDir string) error {
...
+module github.com/pulumi/pulumi-azure-native-sdk
+{{ else }}
+module github.com/pulumi/pulumi-azure-native-sdk/{{ .SubmoduleName }}
+{{ end }}
+
+go 1.17
+
+require (
+	github.com/blang/semver v3.5.1+incompatible
+	github.com/pkg/errors v0.9.1
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/cmd/pulumi-gen-azure-native/main.go:323]: No idea on what we actually use. This is taken from our currently manually committed go.mod in the `sdk` folder which we currently manually maintain. I can add something to the contributing docs around updating this. Long term, I think codegen should be generating these dependencies - so they can be upgraded based on what the generated code will need - similar in approach to [this issue around Python dependencies](https://github.com/pulumi/pulumi/issues/10565)

[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/cmd/pulumi-gen-azure-native/main.go:323]: Just checked and apparently we do actually use that package in the generated SDK.

**Comment by danielrbradley** [comment]
> Should/can we add an acceptance test for this?

So, not entirely straightforward to figure out, but it seems like it should work!

**Comment by danielrbradley** [comment]
@viveklak doesn't look like program tests are likely to work at the time. Worth just doing a quiet release alongside the old SDK so we can continue testing against an actual real release?

**Review by viveklak — APPROVED** [review]

**Pull request merged**